File: src/python/_0841_Keys_and_Rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        seen = set()
        stack = [0]

        while stack:
            roomNumber = stack.pop()

            if roomNumber not in seen:
                for key in rooms[roomNumber]:
                    stack.append(key)

            seen.add(roomNumber)

        return len(seen) == len(rooms)

File: src/java/_0841_Keys_and_Rooms.java
class Solution {
    private void dfs(List<List<Integer>> rooms, boolean[] visited, int idx) {
        if (visited[idx])
            return;

        visited[idx] = true;

        for (int i = 0; i < rooms.get(idx).size(); i++)
            dfs(rooms, visited, rooms.get(idx).get(i));
    }

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];

        dfs(rooms, visited, 0);

        for (int i = 0; i < visited.length; i++)
            if (!visited[i])
                return false;

        return true;
    }
}
